### PR TITLE
YAML ban system

### PR DESCRIPTION
**Pull request by FaigerSYS**
Add YAML ban system. Implemented switcher and updater from default to YAML

### CODE REVIEW

**Comment by ishitatsuyuki** [comment]
Please review the contribution guidelines before opening a pull request.

[Contribution guidelines](https://github.com/iTXTech/Genisys/blob/master/CONTRIBUTING.md)
> **Do not consider spamming pull requests.**

**Comment by FaigerSYS** [comment]
@ishitatsuyuki , ummm, I closed previous PR because of it was "dirty". I rewrited YAML system, and published it there. So please reopen this PR

**Comment by SOF3** [comment]
Multiple ban list formats existing at the same time is definitely a bad idea. Deprecate the old one instead of making an option.

**Comment by FaigerSYS** [comment]
@SOF3 , it's not multiple, only one works at the same time

**Comment by SOF3** [comment]
@FaigerSYS it is still multiple in the same version. It is a bad idea to support multiple formats of a config file in the same version, as it can easily to confusion between the two in the future, and there being no automatic interconversion.

**Comment by FaigerSYS** [comment]
@SOF3 , I would like to rewrite, but PR is closed

File: src/pocketmine/permission/BanList.php [hunk 1]
```diff
@@ -155,13 +152,41 @@ public function save($flag = true){
...
 			}
+
+			fwrite($fp, yaml_emit($data));
 			fclose($fp);
 		}else{
 			MainLogger::getLogger()->error("Could not save ban list");
 		}
 	}
 
+	public static function __toYaml($old_file, $new_file) {
```
Comment (SOF3) [inline src/pocketmine/permission/BanList.php]: Avoid naming methods starting with `__`. They may clash with PHP magic methods.

File: src/pocketmine/permission/BanList.php [hunk 2]
```diff
@@ -37,6 +37,7 @@ class BanList{
 
 	/**
 	 * @param string $file
+	 * @param bool $isYaml
```
Comment (SOF3) [inline src/pocketmine/permission/BanList.php]: This line can be removed.

File: src/pocketmine/permission/BanEntry.php [hunk 3]
```diff
@@ -127,4 +122,5 @@ public static function fromString($str){
 			return $entry;
 		}
 	}
+
```
Comment (SOF3) [inline src/pocketmine/permission/BanEntry.php]: Unnecessary change

File: src/pocketmine/permission/BanEntry.php [hunk 4]
```diff
@@ -82,43 +82,38 @@ public function setReason($reason){
...
+		return $data;
 	}
 
 	/**
-	 * @param string $str
+	 * @param array $data
 	 *
 	 * @return BanEntry
 	 */
-	public static function fromString($str){
```
Comment (SOF3) [inline src/pocketmine/permission/BanEntry.php]: This is not the proper way to deprecate a method. You should remain basic support for the method, and mark it as deprecated.

File: src/pocketmine/permission/BanList.php (hunk 1 above)
Comment (dktapps) [inline src/pocketmine/permission/BanList.php]: The __ is reserved for use with magic methods. This is not a magic method.

File: src/pocketmine/permission/BanEntry.php [hunk 5]
```diff
@@ -82,43 +82,38 @@ public function setReason($reason){
...
-		if(strlen($str) < 2){
+	public static function fromArray(array $data){
+		if(count($data) < 1){
```
Comment (SOF3) [inline src/pocketmine/permission/BanEntry.php]: why not an empty BanList?

File: src/pocketmine/permission/BanEntry.php [hunk 6]
```diff
@@ -82,43 +82,38 @@ public function setReason($reason){
 		$this->reason = $reason;
 	}
 
-	public function getString(){
```
Comment (SOF3) [inline src/pocketmine/permission/BanEntry.php]: This method should be marked as deprecated instead of being removed.

**Comment by dktapps** [comment]
This pull request does not make effective use of the YAML format. It produces a format like:
```
---
- - dktapps
  - "2016-10-14 17:14:47 +0800"
  - CONSOLE
  - Forever
  - forever
...
```
This is a very inefficient way to store such data, and does not utilise the useful features that YAML provides.

**Comment by FaigerSYS** [comment]
@dktapps , what do you suggest? The main thing that there were no errors

**Comment by SOF3** [comment]
@FaigerSYS if we just want no errors, don't use YAML. Just serialize all the stuff and that gets done.

We want a ban list that is easy to read and edit for both programs and users.

**Comment by dktapps** [comment]
I think I'll come up with something myself. This is far too much palaver for something like this, and it _was_ my idea to begin with.

**Comment by FaigerSYS** [comment]
@SOF3 , so you say that format is not easy to read and edit by users.....?
